lib/azure/armrest/configuration.rb

```diff
@@ -106,31 +105,47 @@ def initialize(args)
         user_token = options.delete(:token)
         user_token_expiration = options.delete(:token_expiration)
 
+        # We need to ensure these are set before subscription_id=
+        @tenant_id = options.delete(:tenant_id)
+        @client_id = options.delete(:client_id)
+        @client_key = options.delete(:client_key)
+
+        # Then set the remaining options automatically
         options.each { |key, value| send("#{key}=", value) }
 
-        unless client_id && client_key && tenant_id && subscription_id
-          raise ArgumentError, "client_id, client_key, tenant_id and subscription_id must all be specified"
+        unless client_id && client_key && tenant_id
+          raise ArgumentError, "client_id, client_key, and tenant_id must all be specified"
         end
 
-        # Allows for URI objects or Strings.
-        @proxy = @proxy.to_s if @proxy
-
-        validate_subscription
-
         if user_token && user_token_expiration
           set_token(user_token, user_token_expiration)
         elsif user_token || user_token_expiration
           raise ArgumentError, "token and token_expiration must be both specified"
         end
 
-        @providers = fetch_providers
-        set_provider_api_versions
+        if subscription_id
```

Comment:
Hm, I feel like that's overworking the subscription_id method. I don't think the provider information changes based on the subscription id, so it would potentially be making an unnecessary call.